### PR TITLE
`SharedDirectoryAcknowledge`

### PR DESCRIPTION
**Pull request by ibeckermayer**
Adds `SharedDirectoryAcknowledge` handling and does some tidying up in the client and codec.

~base branch should be changed to master once https://github.com/gravitational/webapps/pull/960 is merged~

requires backport to v9/v10

#### `teleport`
Build and run the `windows-desktop-directory-sharing` branch, editing [this line](https://github.com/gravitational/teleport/blob/15439227ea32be084cf5723ceed0aae72e79dc7a/Makefile#L232) in the Makefile to have the `directory_sharing` build flag:
```
$(BUILDDIR)/teleport: ensure-webassets bpf-bytecode rdpclient
	GOOS=$(OS) GOARCH=$(ARCH) $(CGOFLAG) go build -tags "$(PAM_TAG) $(FIPS_TAG) $(BPF_TAG) $(WEBASSETS_TAG) $(RDPCLIENT_TAG) directory_sharing" -o $(BUILDDIR)/teleport $(BUILDFLAGS) ./tool/teleport
```

#### `webapps`
Change the value below to `true` and run the dev server
https://github.com/gravitational/webapps/blob/becfad38776a68e0d8be634476f24cead7890ffc/packages/teleport/src/config.ts#L32

### What to expect

The teleport logs will have something like
```
[2022-07-18T17:31:04Z DEBUG rdp_client::rdpdr] sending new drive for redirection over RDP: ClientDeviceListAnnounceRequest { device_count: 1, device_list: [DeviceAnnounceHeader { device_type: RDPDR_DTYP_FILESYSTEM, device_id: 2, preferred_dos_name: "webapps", device_data_length: 8, device_data: [119, 101, 98, 97, 112, 112, 115, 0] }] }
[2022-07-18T17:31:04Z DEBUG rdp_client::rdpdr] received RDP: ServerDeviceAnnounceResponse { device_id: 2, result_code: 0 }
[2022-07-18T17:31:04Z DEBUG rdp_client::rdpdr] ServerDeviceAnnounceResponse for shared directory succeeded
[2022-07-18T17:31:04Z DEBUG rdp_client] sending TDP SharedDirectoryAcknowledge: SharedDirectoryAcknowledge { err_code: Nil, directory_id: 2 }
[2022-07-18T17:31:04Z DEBUG rdp_client::rdpdr] received RDP: DeviceCreateRequest { device_io_request: DeviceIoRequest { device_id: 2, file_id: 0, completion_id: 2, major_function: IRP_MJ_CREATE, minor_function: IRP_MN_NONE }, desired_access: FILE_READ_ATTRIBUTES, allocation_size: 0, file_attributes: (empty), shared_access: FILE_SHARE_READ | FILE_SHARE_WRITE | FILE_SHARE_DELETE, create_disposition: FILE_OPEN, create_options: FILE_DIRECTORY_FILE, path_length: 0, path: WindowsPath { path: "" } }
[2022-07-18T17:31:04Z DEBUG rdp_client] sending TDP SharedDirectoryInfoRequest: SharedDirectoryInfoRequest { completion_id: 2, directory_id: 2, path: UnixPath { path: "" } }
```
and the session should fail with the message

image

This means that the backend handled the TDP `SharedDirectoryAnnounce` and `SharedDirectoryAcknowledge` correctly, and started handling an RDP `DeviceCreateRequest` by sending us a TDP `SharedDirectoryInfoRequest` (message type 13), which we don't yet handle.

### CODE REVIEW

File: packages/teleport/src/lib/tdp/codec.test.ts [hunk 1]
```diff
@@ -104,7 +104,9 @@ test('decodes message types', () => {
   const { buffer: pngFrameBuf } = makeBuf(MessageType.PNG_FRAME);
   const { buffer: clipboardBuf } = makeBuf(MessageType.CLIPBOARD_DATA);
   const { buffer: errorBuf } = makeBuf(MessageType.ERROR);
-  const { buffer: invalidBuf } = makeBuf(MessageType.MFA_JSON + 1);
+  const { buffer: invalidBuf } = makeBuf(
```
Comment (hatched) [inline packages/teleport/src/lib/tdp/codec.test.ts:107]: I noticed this is different in another PR so just a note to make sure they land in order to get the correct implementation for the `invalidBuf`

**Review by hatched — APPROVED** [review]

**Comment by ravicious** [comment]
I keep running into "invalid message type: 13" when trying this out on your server. Before I was able to get to this point I got that "Document is not focused" error two or three times.

I changed `enableDirectorySharing` to `true` and ran `yarn start-teleport`, is there anything else I need to do?

https://user-images.githubusercontent.com/27113/179237904-2c16ca87-b085-442b-8d77-c2bd408d90b2.mov

**Comment by kimlisa** [comment]
> I keep running into "invalid message type: 13" when trying this out on your server. Before I was able to get to this point I got that "Document is not focused" error two or three times.

@ravicious i think this PR only handles message type 12

File: packages/teleport/src/lib/tdp/codec.ts [hunk 2]
```diff
@@ -480,13 +505,27 @@ export default class Codec {
...
 
-  // _asBase64Url creates a data:image uri from the png data part of a PNG_FRAME tdp message.
-  _asBase64Url(buffer: ArrayBuffer): string {
+  // | message type (12) | directory_id uint32 | succeeded byte |
+  decodeSharedDirectoryAcknowledge(
+    buffer: ArrayBuffer
+  ): SharedDirectoryAcknowledge {
+    let dv = new DataView(buffer);
+    let errCode = toSharedDirectoryErrCode(dv.getUint32(1));
+    let directoryId = dv.getUint32(5);
```
[the remark this answers is not in the document]
Comment (ibeckermayer) [reply packages/teleport/src/lib/tdp/codec.ts:519]: Replied to a similar question here: https://github.com/gravitational/webapps/pull/966#discussion_r923632090

I will fix this one in this PR and the rest in that one.

**Pull request merged**